traits/tests/test_property_notifications.py

```diff
@@ -45,3 +66,437 @@ def test_property_notifications(self):
+    """ Dummy class to test property observing container items"""
+
+    discounted = Property(
+        Bool(), observes=trait("list_of_infos").list_items().trait("age"))
+
+    discounted_n_calculations = Int()
+
+    def _get_discounted(self):
+        self.discounted_n_calculations += 1
+        return any(info.age > 70 for info in self.list_of_infos)
+
+
+class ClassWithPropertyDependsOnItems(ClassWithInstanceDefaultInit):
+    """ Dummy class using depends_on to be compared with the one using
+    observes."""
+
+    discounted = Property(Bool(), depends_on="list_of_infos.age")
+
+    discounted_n_calculations = Int()
+
+    def _get_discounted(self):
+        self.discounted_n_calculations += 1
+        return any(info.age > 70 for info in self.list_of_infos)
+
+
+class ClassWithPropertyObservesWithCache(PersonInfo):
+
+    discounted = Property(Bool(), observes="age")
+
+    discounted_n_calculations = Int()
+
+    @cached_property
+    def _get_discounted(self):
+        self.discounted_n_calculations += 1
+        return self.age > 10
+
+
+class ClassWithPropertyObservesDecorated(PersonInfo):
+    """ Dummy class to test property with observers setup at init time."""
+
+    discounted = Property(Bool(), observes="age")
+
+    discounted_n_calculations = Int()
+
+    def _get_discounted(self):
+        self.discounted_n_calculations += 1
+        return self.age > 60
+
+    discounted_events = List()
+
+    @observe("discounted")
+    def discounted_updated(self, event):
+        self.discounted_events.append(event)
+
+
+class ClassWithPropertyDependsOnDecorated(PersonInfo):
+    """ Dummy class to test property with depends_on and on_trait_change
+    setup at init time."""
+
+    discounted = Property(Bool(), depends_on="age")
+
+    discounted_n_calculations = Int()
+
+    def _get_discounted(self):
+        self.discounted_n_calculations += 1
+        return self.age > 60
+
+    discounted_events = List()
+
+    @on_trait_change("discounted")
+    def discounted_updated(self, event):
+        self.discounted_events.append(event)
+
+
+class ClassWithPropertyMissingGetter(PersonInfo):
+    """ Dummy class to test property that does not have getter/setter.
+    """
+
+    discounted = Property(Bool(), observes="age")
+
+
+class ClassWithPropertyTraitNotFound(HasTraits):
+    """ Dummy class to test error, prevent issues like enthought/traits#447
+    """
+
+    person = Instance(PersonInfo)
+
+    last_name = Property(observes="person.last_name")  # last_name not defined
+
+
+class TestHasTraitsPropertyObserves(unittest.TestCase):
+    """ Tests Property notifications using 'observes', and compared the
+    behavior with 'depends_on'
+    """
+
+    def setUp(self):
+        push_exception_handler(reraise_exceptions=True)
+        self.addCleanup(pop_exception_handler)
+
+    def test_property_observes_extended_trait(self):
+        instance_observes = ClassWithPropertyObservesDefault()
+        handler_observes = mock.Mock()
```

Comment:
@mdickinson These examples were written before #1208 was reviewed. Would you like me to change these mock objects to `list.append`?

Reply:
I'm fine either way: it's just in tests that intended to be integration tests that I'd like to avoid use of `mock`.